registry-library/library/library_test.go

```diff
@@ -239,10 +327,385 @@ func TestGetRegistryIndex(t *testing.T) {
+			if err != nil {
+				t.Errorf("Unexpected error while matching url: %v", err)
+				return
+			}
+
+			buffer := bytespkg.Buffer{}
+			writer := zip.NewWriter(&buffer)
+
+			_, err = writer.Create("README.md")
+			if err != nil {
+				t.Errorf("error in creating testing starter project archive: %v", err)
+				return
+			}
+
+			writer.Close()
+
+			bytes = buffer.Bytes()
+		} else if strings.HasPrefix(r.URL.Path, "/index") || strings.HasPrefix(r.URL.Path, "/v2index") {
+			data := setUpIndexHandle(r.URL)
+
+			bytes, err = json.MarshalIndent(&data, "", "  ")
+			if err != nil {
+				t.Errorf("Unexpected error while doing json marshal: %v", err)
+				return
+			}
+		} else {
+			t.Errorf("Route %s was not found", r.URL.Path)
+			return
+		}
+
+		_, err = w.Write(bytes)
+		if err != nil {
+			t.Errorf("Unexpected error while writing data: %v", err)
+		}
+	})
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer close()
+
+	tests := []struct {
+		name           string
+		url            string
+		stack          string
+		starterProject string
+		options        RegistryOptions
+		wantType       string
+		wantErr        bool
+	}{
+		{
+			name:           "Download Starter Project",
+			url:            "http://" + serverIP,
+			stack:          "stackindex1",
+			starterProject: "stackindex1-starter",
+			wantType:       "application/zip",
+		},
+		{
+			name:           "Download Starter Project V2",
+			url:            "http://" + serverIP,
+			stack:          "stackv2index1",
+			starterProject: "stackv2index1-starter",
+			options: RegistryOptions{
+				NewIndexSchema: true,
+			},
+			wantType: "application/zip",
+		},
+		{
+			name:           "Download Starter Project from Fake Stack",
+			url:            "http://" + serverIP,
+			stack:          "fakestack",
+			starterProject: "fakestarter",
+			wantErr:        true,
+		},
+		{
+			name:           "Download Fake Starter Project",
+			url:            "http://" + serverIP,
+			stack:          "stackindex1",
+			starterProject: "fakestarter",
+			wantErr:        true,
+		},
+		{
+			name:           "Download Fake Starter Project V2",
+			url:            "http://" + serverIP,
+			stack:          "stackv2index1",
+			starterProject: "fakestarter",
+			options: RegistryOptions{
+				NewIndexSchema: true,
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			gotBytes, err := DownloadStarterProjectAsBytes(test.url, test.stack, test.starterProject, test.options)
+
+			if !test.wantErr && err != nil {
+				t.Errorf("Unexpected err: %+v", err)
+			} else if (test.wantErr || gotBytes == nil) && err == nil {
+				t.Error("Expected error but got nil")
+			} else if test.wantType != "" {
+				gotType := http.DetectContentType(gotBytes)
+				if !reflect.DeepEqual(gotType, test.wantType) {
+					t.Errorf("Expected: %+v, \nGot: %+v", test.wantType, gotType)
+				}
+			}
+		})
+	}
+}
```

Comment:
Do we have tests for `DownloadStarterProjectAsDir` and `DownloadStarterProject` ?

Reply:
@kim-tsao Missed these, the test functions are now added.